src/box/gc.c

```diff
@@ -733,6 +806,280 @@ gc_consumer_name(struct gc_consumer *consumer)
 			  tt_uuid_str(&consumer->uuid));
 }
 
+/**
+ * Write the consumer to persistent WAL GC state. The function must be called
+ * only when schema supports persistent consumers. If consumer wasn't in space
+ * _gc_consumers, the new row is created.
+ */
+static int
+gc_consumer_persist_impl(struct gc_consumer *consumer)
+{
+	assert(gc_schema_supports_persistent_consumers());
+
+	/* Mark consumer as synchronized with persistent state. */
+	consumer->is_synced = true;
+	char tuple_buf[UUID_STR_LEN + VCLOCK_STR_LEN_MAX + GC_NAME_MAX + 30];
+	char *tuple_end = tuple_buf;
+	tuple_end = mp_encode_array(tuple_end, 3);
+	tuple_end = mp_encode_str0(tuple_end, tt_uuid_str(&consumer->uuid));
+	tuple_end = mp_encode_vclock_ignore0(tuple_end, &consumer->vclock);
+	tuple_end = mp_encode_map(tuple_end, 1);
+	tuple_end = mp_encode_str0(tuple_end, "type");
+	tuple_end = mp_encode_str0(tuple_end,
+				   gc_consumer_type_strs[consumer->type]);
+	assert((unsigned long)(tuple_end - tuple_buf) < sizeof(tuple_buf));
+
+	return box_replace(BOX_GC_CONSUMERS_ID, tuple_buf, tuple_end, NULL);
+}
+
+int
+gc_consumer_persist(struct gc_consumer *consumer)
+{
+	if (gc_schema_supports_persistent_consumers()) {
+		struct txn *txn = txn_begin();
+		if (txn == NULL)
+			return -1;
+		/*
+		 * No need to wait limbo, so let's set TXN_FORCE_ASYNC.
+		 * Moreover, the function can be called when quorum cannot be
+		 * collected, (for example, on IPROTO_SUBSCRIBE), so we must set
+		 * this flag here.
+		 */
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (gc_consumer_persist_impl(consumer) != 0) {
+			txn_abort(txn);
+			return -1;
+		}
+		if (txn_commit(txn) != 0)
+			return -1;
+	}
+	/* Unref WAL GC delay if being persisted for the first time. */
+	if (!consumer->is_persistent)
+		gc_delay_unref();
+	consumer->is_persistent = true;
+	return 0;
+}
+
+/**
+ * Mark consumer as not persistent and unregister it if unused.
+ */
+static int
+gc_erase_consumer_on_commit(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct gc_consumer *consumer = trigger->data;
+	if (consumer != NULL) {
+		consumer->is_persistent = false;
+		if (consumer->is_orphan)
+			gc_consumer_unregister_impl(consumer);
+	}
+	return 0;
+}
+
+int
+gc_erase_consumer(const struct tt_uuid *uuid)
+{
+	assert(in_txn() != NULL);
+	struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+	if (!gc_schema_supports_persistent_consumers()) {
+		consumer->is_persistent = false;
+		return 0;
+	}
+
+	if (boxk(IPROTO_DELETE, BOX_GC_CONSUMERS_ID, "[%s]",
+		 tt_uuid_str(uuid)) != 0)
+		return -1;
+
+	/* Do actual work on commit. */
+	struct trigger *on_commit =
+		xregion_alloc_object(&in_txn()->region, struct trigger);
+	trigger_create(on_commit, gc_erase_consumer_on_commit, consumer, NULL);
+	txn_stmt_on_commit(txn_current_stmt(in_txn()), on_commit);
+	return 0;
+}
```

Comment:
Personally, I'd prefer if this function worked the other way around:
1. Clear the `is_persistent` flag.
2. Delete the consumer from the space.
3. Install a commit trigger that would drop the in-memory consumer object and a rollback trigger that would set back the `is_persistent` flag.

This way we wouldn't need to have two code paths for updating the consumers space (`gc_consumer_sync_impl` and `gc_consumer_persist_impl`), neither would we need to write all consumers to the space on the schema version update (`gc_persist_consumers`).

I don't insist though since there may be other issues with this approach. Leaving this up to @Gerold103 and @drewdzzz to decide.

Reply:
Patch is already approved with current approach  two code paths for updating the consumers space), so I resolve the comment.